### DISCUSSION TITLE
How to enable timestamping and LTV?

### DISCUSSION THREAD

**Discussion by agunghp** [gettalong/hexapdf#201] [Q&A] [answered]
Hi, is there any way to enable timestamping and LTV using HexaPDF? 
Thanks

**Comment by gettalong** [comment]
No, this is not possible at the moment. However, this is on my TODO list and I'm currently looking into adding support for both.

**Reply by agunghp** [reply]
Got it. Looking forward for the support

**Comment by gettalong** [comment, marked as answer]
It is now possible to create PAdES level B-B and B-T)signatures and the latter contain timestamps from a trusted timestamping server. Old-style CMS signatures with a timestamp are also possible.

Those signatures appear as "LTV enabled" in certain viewers but are not LTV.

So this answers part of your question.

**Reply by agunghp** [reply]
Thank you for the update this is very helpful. 

It will be great if it can fully support LTV in the future.

**Comment by maxfindel** [comment]
Hey @gettalong thanks for this update. The support for PAdES is working great for me. I was looking into adding the DSS information required to have LTV and I noticed that it's added as a sibling Dictionary next to AcroForm, Metadata and Pages as the images show:

<img width="408" alt="Screen Shot 2023-06-17 at 18 00 55" src="https://github.com/gettalong/hexapdf/assets/888004/d5298c00-c5e0-4934-a70d-6eedfd7a8cb0">

<img width="391" alt="Screen Shot 2023-06-17 at 18 00 49" src="https://github.com/gettalong/hexapdf/assets/888004/8dd9180c-814b-4ddc-aa62-43e813ae9330">

To achieve this, I used the "Add Verification Info" option inside Acrobat Reader on the signed PDF and the result was a modified file that has LTV. 

To get the CRL and OCSP responses [I tried this approach](https://ruby-doc.org/core-3.1.2/OpenSSL/OCSP.html) and it worked with my certs (which have CRL and OCSP endpoints). I was able to get a valid response, which can be later added to the PDF.

I think this code could be added to hexapdf as a standalone validation class. This validation can be done after the signature process and then the result can be added to the right dictionary, constructing the data similarly to the `signed_data_creator.rb` class.

Can you point me in the right direction in order to add the DSS data to the PDF catalog? Thanks!

**Reply by gettalong** [reply]
The DSS dictionary is part of the PDF 2.0 specification which can now be retrieved for free, see https://www.pdfa-inc.org/product/iso-32000-2-pdf-2-0-bundle-sponsored-access/. There your will find it section 12.8.4 "Longterm validation of signatures".

In HexaPDF you would access or modify the DSS dictionary via `document.catalog[:DSS]` (though I still need to update the existing type classes like `HexaPDF::Type::Catalog` to reflect the additional PDF 2.0 fields). So you could do:

~~~ ruby
dss = {Certs: [cert1, cert2, cert3], OCSPs: [ocsp_response]}
document.catalog[:DSS] = dss
~~~

Make sure that you use incremental writing when saving the modified document to ensure the validness of the signature.

**Reply by maxfindel** [reply]
Thanks for the tips! I'll try adding the validation manually to the catalog and see how it goes 🙌

**Reply by gettalong** [reply]
The devil is, as usual, in the details but the PDF 2.0 spec should provide enough information to make it work.

As written before, this is on my TODO list but I'm currently working on another part of HexaPDF.

**Reply by maxfindel** [reply]
I read the ISO standard as you suggested and was able to replicate what Adobe Reader does with the "Add Verification Information" with a signature done in HexaPDF 👏 
![Screen Shot 2023-06-23 at 15 00 09](https://github.com/gettalong/hexapdf/assets/888004/b43e215f-168d-4791-bfda-c77d757edc87)

The code is a bit long and not so elegant as the rest of the library, but here it is in case you want to use it, Thomas:
https://gist.github.com/maxfindel/af4afccc6261d7053946dbe2dd837b14

I did encounter an issue that I think is worth mentioning:
- Adobe Reader (and possibly other readers) store cached versions of the CRL and OCSP responses. Thus, I could select the "Add Verification Information" on a signature that used a certificate that was already expired. 
- The issue arised when I was doing the validation using ruby but the end result was not LTV enabled! I found this very frustrating, but in the end it was just a date error, since I was writing the code days after I signed the document.
- This problem should not occur if you perform the validation right after the signature is performed, even if the certificate is valid for, say, 5 minutes. 
- If it helps anyone in the future, here's the decoded ASN1 that shows the validity window of a CRL response (if your signature time is not inside this window, you can't use it as validation):
![Screen Shot 2023-06-23 at 14 11 24](https://github.com/gettalong/hexapdf/assets/888004/6b06ef0e-8573-4dee-bde5-bddf266e8e0a)

**Reply by gettalong** [reply]
Thanks! That looks great! Glad that you got it working!

**Reply by dmke** [reply]
Hi!

I'm having trouble reproducing a PAdES B-LTV signature with this. At best, Acrobat reports that the

> Signer's identity could not be verified because of errors during validation
>
> <details><summary>Screenshot</summary>
> 
> ![Screenshot](https://github.com/gettalong/hexapdf/assets/327411/38010421-223a-416c-a3d3-9059aa905e78))
> 
> </details>

while declaring the signature level is B-B.

The [ETSI conformance checker](https://signatures-conformance-checker.etsi.org) reports no structural errors, and the [signature check on RTR.at](https://signaturpruefung.gv.at) attests:

> The document is signed correctly. However, the signature/seal does not cover the entire document.

(which is not surprising, as the DSS data didn't exist when the signature was produced).

I have published the code and reports at https://github.com/digineo/hexapdf-pades-ltv-repro (the interesting bits live in the `sign.rb` file).

@gettalong, would you mind having a look? Maybe you can find some obvious errors (something about the forest and trees :))

**Reply by gettalong** [reply]
@dmke Sure, I will have time later to have a look at it.

**Reply by gettalong** [reply]
@dmke Thanks for providing the repository and the code!

Some observations:

* When I run `hexapdf info documents/with_signature.pdf` it reports "Error:          Certificate key usage is missing 'Digital Signature'". Looking at the used certificate it only shows "Non-Repudiation" as allowed "Key Usage". So that might be a problem.

* Looking at the signature object using `openssl cms -in /tmp/sig -inform DER -print -cmsout`, I *think* it looks okay.

* I get an error in Adobe Acrobat already for `with_signature.pdf`, so I don't think this has anything to do with DSS and making the signature an LTV one.

Could you try the basic signing with a key that has the 'Digital Signature' key usage?

**Reply by dmke** [reply]
@gettalong Thank you for your help!

I don't think the missing `digitalSignature` key usage bit is relevant here, as we technically/legally don't want to *sign* the document, but *seal* it (as in "Qualified Electronic Signatures", eIDAS Art. 3).

Using the [DSS Demo WebApp on ec.europa.eu](https://ec.europa.eu/digital-building-blocks/DSS/webapp-demo/validation), I have found a hint as to why both the `with_signature.pdf` and `with_dss.pdf` fail to validate:

<details><summary><code>with_signature.pdf</code></summary>

![image](https://github.com/gettalong/hexapdf/assets/327411/c495b139-52cf-4f9d-812f-44de7e3eb8cf)

</details>
<details><summary><code>with_dss.pdf</code></summary>

![image](https://github.com/gettalong/hexapdf/assets/327411/c2c04337-501b-4f7f-a5f6-8e7de5e78c51)

</details>
<details><summary>(N.B. that tool has an exhaustive "detailed report")</summary>

![image](https://github.com/gettalong/hexapdf/assets/327411/3da83eb8-1805-4302-ad72-e68034c139e7)

</details>

|   | `with_signature.pdf` | `with_dss.pdf` |
|:--|:---------------------|:---------------|
| Qualification | QESeal | *same* |
| Signature format | PAdES-BASELINE-T | PAdES-BASELINE-LT |
| Indication | TOTAL_PASSED | *same* |
| AdES Validation Details | **The signed attribute: 'signing-certificate' is absent!** | *same* |

It seems I need to amend the PKCS#7 structure in our signing service a bit.

I'll let you know how I'll progress.


---

Edit: Cool, the DSS Demo WebApp is Open Source: https://github.com/esig

**Reply by dmke** [reply]
> The signed attribute: 'signing-certificate' is absent!

I believe this means *ESS signing-certificate* ([CAdES 5.2.2.2](https://www.etsi.org/deliver/etsi_en/319100_319199/31912201/01.02.01_60/en_31912201v010201p.pdf), as opposed to *ESS signing-certificate-v2*, 5.2.2.3). Those are mutually exclusive...

**Reply by gettalong** [reply]
@dmke Yeah, the web-based tools from the EU are quite good for checking PDFs!

One note: While the CMS signed data structure in `with_signature.pdf` doesn't have the 'signing-certificate', it has the Version 2 of it:

~~~
    signerInfos:
        version: 1
        d.issuerAndSerialNumber:
          issuer: C=EE, O=SK ID Solutions AS/organizationIdentifier=NTREE-10747013, CN=SK ID Solutions ORG 2021E
          serialNumber: 115411033961928712430093513817117753433
        digestAlgorithm:
          algorithm: sha256 (2.16.840.1.101.3.4.2.1)
          parameter: <ABSENT>
        signedAttrs:
            object: contentType (1.2.840.113549.1.9.3)
            set:
              OBJECT:pkcs7-data (1.2.840.113549.1.7.1)

            object: messageDigest (1.2.840.113549.1.9.4)
            set:
              OCTET STRING:
                0000 - 50 c9 0f af 80 99 e0 8c-94 4b 35 e6 3d   P........K5.=
                000d - a9 75 2e b1 06 93 ea d7-59 60 04 ed a2   .u......Y`...
                001a - 01 d8 2a 69 f3 da                        ..*i..

            object: id-smime-aa-signingCertificateV2 (1.2.840.113549.1.9.16.2.47)
...
~~~

The standard ETSI EN 319 122-1 V1.1.1 section 5.2.2 specifies that either v1 or v2 must be present, also see ETSI EN 319 142-1 V1.1.1 table 1 on page 19 and page 21 where it says: "Generators shall use either the signing certificate or the signing-certificate v2 attribute, depending on the hash function, in accordance with ETSI EN 319 122-1 [2]".

Maybe related:

https://github.com/gettalong/hexapdf/blob/e784ea875a60e04512bcbb28fd99c79044ad9941/lib/hexapdf/digital_signature/signing/signed_data_creator.rb#L134-L135

**Reply by dmke** [reply]
I'm trying to avoid SHA-1 wherever possible, so `id-aa-signingCertificateV2` is a requirement.

> Maybe related:
>
> ```rb
> #TODO: Does not validate on ETSI checker if used, doesn't matter if SHA256 or 512 
> #oid('sha512')
> ```

The signing library we've based our service on only includes the ESSCertID(v2) AlgorithmIdentifier, if the hash function is neither SHA-1 nor SHA-256:

https://github.com/digitorus/pdfsign/blob/e6e59033a046f4649f131bcffcf680b831510c15/sign/pdfsignature.go#L135-L139

(The main difference seems to be that the AlgorithmIdentifier is wrapped in a sequence, which HexaPDF omits (if I read the code correctly).)

I'm going to try [PLOP DS](https://www.pdflib.com/de/produkte/plop-ds/funktionalitaet/) next. Maybe the evaluation license allows to generate a signature/seal with a more widely recognized validity, so I can spot the difference in the PDF and/or CMS structure. If that fails, I'll annoy Adobe :)

**Reply by gettalong** [reply]
What is missing from HexaPDF's signing feature that doesn't allow you to use it? I.e. so that you wouldn't need to depend on a third-party tool for creating the CMS structure. Signing the hash would still need to be done externally - I guess - since you are creating a qualified signature where you need a secure device to sign the hash.

**Reply by dmke** [reply]
I had started with a `DefaultHandler` subclass (overriding the `#sign` method), but I'm not sure anymore why I haven't followed through.

IIRC, the `SignedDataCreator` required the certificate chain to already be present in `#create_signed_attrs`, but we could only extract it in the `#digest_and_sign_data` call (which happened afterwards). There were also multiple warnings and errors in the ETSI conformance checker (but those might have been due to user error).

Thinking about it, it might be sensible to go back to a `DefaultHandler` subclass. Since the QSCD certificate chain is relatively static, our service can expose them on a separate endpoint, so that the sealer code could cache them locally. On the other hand, it *feels* like there's not much missing here—but that's something I've told myself (and my boss) on every turn so far :)

**Reply by gettalong** [reply]
The method `#digest_and_sign_data` is called before `#create_signed_data` where the certificates are embedded. So setting them in `#digest_and_sign_data` should work. Only the signing certificate itself is needed before `#digest_and_sign_data`.

**Reply by gettalong** [reply]
> IIRC, the SignedDataCreator required the certificate chain to already be present in #create_signed_attrs, but we could only extract it in the #digest_and_sign_data call (which happened afterwards). There were also multiple warnings and errors in the ETSI conformance checker (but those might have been due to user error).

If you like, I could have a look at those PDF files and see if I can find something.

**Reply by dmke** [reply]
I don't have access to those PDF files anymore. It was early in the development, and large changes happened in rapid succession.

I'm going to spend a little time tomorrow (well... later) restoring a previous version. If you're interested, I may also be able to provide you access to the signature service, so you don't need to play with fixtures.

**Comment by funnyphone** [comment]
hi maxfindel

It seems that following link is not available now. Could you please share the source patch again? thanks a lot.

> The code is a bit long and not so elegant as the rest of the library, but here it is in case you want to use it, Thomas:
> https://gist.github.com/maxfindel/af4afccc6261d7053946dbe2dd837b14

**Reply by gettalong** [reply]
I just tried it and it works for me.

**Reply by funnyphone** [reply]
hi Thomas, That's great. Thanks a lot.
